File: py/src/comms/CborEncoder.py
from typing import Type
import cbor

from .Encoder import Encoder
from .jsonRpc import RequestOrResponse

class CborEncoder(Encoder):

    def name(self) -> str:
        return 'cbor'

    def decode(self, message: bytes, cls: Type[RequestOrResponse]) -> RequestOrResponse:
        obj = cls()
        dic = cbor.loads(message)
        obj.__dict__.update(dic)
        return obj

    def encode(self, obj: RequestOrResponse) -> bytes:
        return cbor.dumps(obj.__dict__)

File: js/src/comms/CborEncoder.ts
import cbor from 'cbor'

import Encoder from './Encoder'
import JsonRpcRequest from './JsonRpcRequest'
import JsonRpcResponse from './JsonRpcResponse'

/**
 * A [CBOR](http://cbor.io) message encoder
 */
export default class CborEncoder extends Encoder {

  name (): string {
    return 'cbor'
  }

  encode (instance: JsonRpcRequest | JsonRpcResponse): Uint8Array {
    return cbor.encode(instance)
  }

  decode<Type = JsonRpcRequest | JsonRpcResponse> (message: Uint8Array, constructor: new () => Type): Type {
    const instance = new constructor()
    const data = cbor.decodeFirstSync(Buffer.from(message))
    Object.assign(instance, data)
    return instance
  }
}
